### PR TITLE
Coproduct instances expose ordinal

### PR DESCRIPTION
**Pull request by TimWSpence**
Expose ordinals in `CoproductInstances#fold2` for the sake of eg deriving `Ord`

### CODE REVIEW

File: modules/deriving/src/test/scala/shapeless3/deriving/deriving.scala [hunk 1]
```diff
@@ -78,6 +78,48 @@ class DerivationTests {
...
+    assert(v1.compare(Box(1), Box(1)) == 0)
+    assert(v1.compare(Box(1), Box(0)) == 1)
+    assert(v1.compare(Box(0), Box(1)) == -1)
+
+    val v2 = Ord[OptionInt]
+    assert(v2.compare(NoneInt, NoneInt) == 0)
+    assert(v2.compare(SomeInt(0), SomeInt(0)) == 0)
+    assert(v2.compare(SomeInt(0), SomeInt(1)) == -1)
+    assert(v2.compare(SomeInt(1), SomeInt(0)) == 1)
+    assert(v2.compare(SomeInt(0), NoneInt) == -1)
```
Comment (milessabin) [inline modules/deriving/src/test/scala/shapeless3/deriving/deriving.scala:96]: I guess I find this ordering a bit strange.

**Review by milessabin — APPROVED** [review]
I find the ordinal based ordering a bit arbitrary ... is that really how it's done in Haskell? `Just(1)` < `Nothing`?

Aside from that this all looks great :-)

**Comment by TimWSpence** [comment]
> I find the ordinal based ordering a bit arbitrary ... is that really how it's done in Haskell? `Just(1)` < `Nothing`?
> 
> Aside from that this all looks great :-)

Yeah, they declare the data constructors in the opposite order so you get the more natural ordering. I've done the same here now

File: modules/deriving/src/test/scala/shapeless3/deriving/deriving.scala (hunk 1 above)
Comment (joroKr21) [reply modules/deriving/src/test/scala/shapeless3/deriving/deriving.scala]: It's definitely better than alphabetic ordering 😜

Comment (milessabin) [reply modules/deriving/src/test/scala/shapeless3/deriving/deriving.scala]: :roll_eyes:

**Pull request merged**